src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/ReadyToRunHeader.cs

```diff
@@ -7,13 +7,92 @@
 {
+    /// <summary>
+    /// Structure representing an element of the assembly table in composite R2R images.
+    /// </summary>
+    public class ComponentAssembly
+    {
+        public const int Size = 4 * sizeof(int);
+
+        public readonly int CorHeaderRVA;
+        public readonly int CorHeaderSize;
+        public readonly int AssemblyHeaderRVA;
+        public readonly int AssemblyHeaderSize;
+
+        public ComponentAssembly(byte[] image, ref int curOffset)
+        {
+            CorHeaderRVA = BitConverter.ToInt32(image, curOffset);
+            curOffset += sizeof(int);
+            CorHeaderSize = BitConverter.ToInt32(image, curOffset);
+            curOffset += sizeof(int);
+            AssemblyHeaderRVA = BitConverter.ToInt32(image, curOffset);
+            curOffset += sizeof(int);
+            AssemblyHeaderSize = BitConverter.ToInt32(image, curOffset);
+            curOffset += sizeof(int);
+        }
+    }
+
+    /// <summary>
+    /// Fields common to the global R2R header and per assembly headers in composite R2R images.
+    /// </summary>
+    public class ReadyToRunCoreHeader
+    {
+        /// <summary>
+        /// Flags in the header
+        /// eg. PLATFORM_NEUTRAL_SOURCE, SKIP_TYPE_VALIDATION
+        /// </summary>
+        public uint Flags { get; set; }
+
+        /// <summary>
+        /// The ReadyToRun section RVAs and sizes
+        /// </summary>
+        public IDictionary<ReadyToRunSectionType, ReadyToRunSection> Sections { get; private set; }
+
+        public ReadyToRunCoreHeader()
+        {
+        }
+
+        public ReadyToRunCoreHeader(byte[] image, ref int curOffset)
+        {
+            ParseCoreHeader(image, ref curOffset);
+        }
+
+        /// <summary>
+        /// Parse core header fields common to global R2R file header and per assembly headers in composite R2R images.
+        /// </summary>
+        /// <param name="image">PE image</param>
+        /// <param name="curOffset">Index in the image byte array to the start of the ReadyToRun core header</param>
+        public void ParseCoreHeader(byte[] image, ref int curOffset)
+        {
+            Flags = NativeReader.ReadUInt32(image, ref curOffset);
+            int nSections = NativeReader.ReadInt32(image, ref curOffset);
+            Sections = new Dictionary<ReadyToRunSectionType, ReadyToRunSection>();
+
+            for (int i = 0; i < nSections; i++)
+            {
+                int type = NativeReader.ReadInt32(image, ref curOffset);
+                var sectionType = (ReadyToRunSectionType)type;
+                if (!Enum.IsDefined(typeof(ReadyToRunSectionType), type))
+                {
+                    // TODO (refactoring) - what should we do?
+                    // R2RDump.WriteWarning("Invalid ReadyToRun section type");
+                }
+                Sections[sectionType] = new ReadyToRunSection(sectionType,
+                    NativeReader.ReadInt32(image, ref curOffset),
```

Comment:
A nit - it would be nice to use named parameters here so that it is obvious what do the two int parameters represent.

Reply:
Hmm, this is not good, I copied this over from pre-existing code but I believe we shouldn't be calling two functions with side effects on the curOffset parameters as part of a single expression. I'll rewrite this as multiple lines.